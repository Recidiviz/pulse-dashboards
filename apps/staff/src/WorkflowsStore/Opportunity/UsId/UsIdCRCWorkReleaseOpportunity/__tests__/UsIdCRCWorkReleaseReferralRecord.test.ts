// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  UsIdCRCWorkReleaseReferralRecordRaw,
  usIdCRCWorkReleaseSchema,
} from "../UsIdCRCWorkReleaseReferralRecord";

describe("UsIdCRCWorkReleaseReferralRecord", () => {
  let rawRecord: UsIdCRCWorkReleaseReferralRecordRaw;

  beforeEach(() => {
    rawRecord = {
      stateCode: "US_ID",
      externalId: "crc-work-release-eligible-01",
      eligibleCriteria: {
        custodyLevelIsMinimum: {
          custodyLevel: "MINIMUM",
        },
        notServingForSexualOffense: null,
        usIdNoDetainersForXcrcAndCrc: null,
        usIdNoAbsconsionEscapeAndEludingPoliceOffensesWithin10Years: null,
        usIdCrcWorkReleaseTimeBasedCriteria: {
          reasons: [
            {
              criteriaName:
                "US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_FTCD_OR_TPD",
              fullTermCompletionDate: "2023-10-10",
              tentativeParoleDate: null,
            },
            {
              criteriaName:
                "US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_EPRD_AND_15_YEARS_OF_FTCD",
              fullTermCompletionDate: "2031-03-13",
              minTermCompletionDate: "2023-11-15",
            },
            {
              criteriaName:
                "US_IX_INCARCERATION_WITHIN_1_YEAR_OF_TPD_AND_LIFE_SENTENCE",
              tentativeParoleDate: "2024-08-14",
            },
          ],
        },
      },
      ineligibleCriteria: {},
    };
  });

  test("transforms eligible record with multiple temporal criteria", () => {
    expect(usIdCRCWorkReleaseSchema.parse(rawRecord)).toMatchSnapshot();
  });

  test("US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_EPRD_AND_15_YEARS_OF_FTCD should be `undefined`", () => {
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfEprdAnd15YearsOfFtcd,
    ).toBeUndefined();
  });

  test("US_IX_INCARCERATION_WITHIN_1_YEAR_OF_TPD_AND_LIFE_SENTENCE should be `undefined`", () => {
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfFtcdOrTpd,
    ).not.toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin1YearOfTpdAndLifeSentence,
    ).toBeUndefined();
  });

  test("US_IX_INCARCERATION_WITHIN_1_YEAR_OF_TPD_AND_LIFE_SENTENCE and US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_EPRD_AND_15_YEARS_OF_FTCD should be `undefined`", () => {
    rawRecord.eligibleCriteria.usIdCrcWorkReleaseTimeBasedCriteria.reasons.splice(
      1,
      1,
    );

    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfFtcdOrTpd,
    ).not.toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin1YearOfTpdAndLifeSentence,
    ).toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfEprdAnd15YearsOfFtcd,
    ).toBeUndefined();
  });

  test("US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_FTCD_OR_TPD and US_IX_INCARCERATION_WITHIN_1_YEAR_OF_TPD_AND_LIFE_SENTENCE should be `undefined`", () => {
    rawRecord.eligibleCriteria.usIdCrcWorkReleaseTimeBasedCriteria.reasons.splice(
      0,
      1,
    );

    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfFtcdOrTpd,
    ).toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfEprdAnd15YearsOfFtcd,
    ).not.toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin1YearOfTpdAndLifeSentence,
    ).toBeUndefined();
  });

  test("US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_FTCD_OR_TPD and US_IX_INCARCERATION_WITHIN_18_MONTHS_OF_EPRD_AND_15_YEARS_OF_FTCD should be `undefined`", () => {
    rawRecord.eligibleCriteria.usIdCrcWorkReleaseTimeBasedCriteria.reasons.splice(
      0,
      2,
    );
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfEprdAnd15YearsOfFtcd,
    ).toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin18MonthsOfFtcdOrTpd,
    ).toBeUndefined();
    expect(
      usIdCRCWorkReleaseSchema.parse(rawRecord).eligibleCriteria
        .usIdIncarcerationWithin1YearOfTpdAndLifeSentence,
    ).not.toBeUndefined();
  });
});

test("transforms eligible record with life temporal criteria", () => {
  const rawRecord: UsIdCRCWorkReleaseReferralRecordRaw = {
    stateCode: "US_ID",
    externalId: "crc-work-release-eligible-02",
    eligibleCriteria: {
      custodyLevelIsMinimum: {
        custodyLevel: "MINIMUM",
      },
      notServingForSexualOffense: null,
      usIdNoDetainersForXcrcAndCrc: null,
      usIdNoAbsconsionEscapeAndEludingPoliceOffensesWithin10Years: null,
      usIdCrcWorkReleaseTimeBasedCriteria: {
        reasons: [
          {
            criteriaName:
              "US_IX_INCARCERATION_WITHIN_1_YEAR_OF_TPD_AND_LIFE_SENTENCE",
            tentativeParoleDate: "2024-08-14",
          },
        ],
      },
    },
    ineligibleCriteria: {},
  };

  expect(usIdCRCWorkReleaseSchema.parse(rawRecord)).toMatchSnapshot();
});
