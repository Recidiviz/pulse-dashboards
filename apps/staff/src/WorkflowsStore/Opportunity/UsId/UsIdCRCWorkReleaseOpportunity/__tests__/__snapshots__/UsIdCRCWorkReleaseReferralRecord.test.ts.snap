// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsIdCRCWorkReleaseReferralRecord transforms eligible record with multiple temporal criteria 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "custodyLevelIsMinimum": Object {
      "custodyLevel": "MINIMUM",
    },
    "notServingForSexualOffense": null,
    "usIdIncarcerationWithin18MonthsOfFtcdOrTpd": Object {
      "fullTermCompletionDate": 2023-10-10T00:00:00.000Z,
      "tentativeParoleDate": null,
    },
    "usIdNoAbsconsionEscapeAndEludingPoliceOffensesWithin10Years": null,
    "usIdNoDetainersForXcrcAndCrc": null,
  },
  "externalId": "crc-work-release-eligible-01",
  "ineligibleCriteria": Object {},
  "stateCode": "US_ID",
}
`;

exports[`transforms eligible record with life temporal criteria 1`] = `
Object {
  "caseNotes": Object {},
  "eligibleCriteria": Object {
    "custodyLevelIsMinimum": Object {
      "custodyLevel": "MINIMUM",
    },
    "notServingForSexualOffense": null,
    "usIdIncarcerationWithin1YearOfTpdAndLifeSentence": Object {
      "tentativeParoleDate": 2024-08-14T00:00:00.000Z,
    },
    "usIdNoAbsconsionEscapeAndEludingPoliceOffensesWithin10Years": null,
    "usIdNoDetainersForXcrcAndCrc": null,
  },
  "externalId": "crc-work-release-eligible-02",
  "ineligibleCriteria": Object {},
  "stateCode": "US_ID",
}
`;
